### PR TITLE
fix(kad): compute `jobs_query_capacity` accurately

### LINKED ISSUE #4191: protocols/kad: accounting error for background job capacity
## Summary

In the [kad behaviour poll function](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L2408) the current query capacity is calculated. Then there is a loop over upto `num` possible provider announcement jobs, possibly starting any that are ready. However a soon as one returns pending the loop is exited via break. Then the current query capacity is decreased by `num`, however due to the break the number of queries started may be less than `num`. It seems to me that the capacity should be reduced by the iteration reached in the loop, not the upper bound. (here: https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L2428)

Then the code proceeds to iterate over replication jobs, using the reduced capacity as a limit on the number that may be started.

## Expected behaviour

`jobs_query_capacity` should reflect an accurate capacity for performing replication / publication jobs

## Actual behaviour

`jobs_query_capacity` is always lower than the actual capacity so fewer replication / publication jobs can be run than expected

## Would you like to work on fixing this bug?

Maybe, although I have no actual Rust skills :grin:

### PR DESCRIPTION
**Pull request by junekhan**
## Description

Fixes: #4191.

### CODE REVIEW

**Review by thomaseizinger — COMMENTED** [review]
Thank you!

Can we add a test please to ensure we don't break this in the future?

**Comment by junekhan** [comment]
> Thank you!
> 
> Can we add a test please to ensure we don't break this in the future?

@thomaseizinger Sure thing! But frankly speaking, I'm a novice in Rust, and it may take a good amount of time to finalize the test case. And, can I ask you for a favor to show me the best test case that I can refer to in [this file](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour/test.rs)? I feel it can somehow speed up us a little bit. Thanks!

**Comment by thomaseizinger** [comment]
> > Thank you!
> > Can we add a test please to ensure we don't break this in the future?
> 
> @thomaseizinger Sure thing! But frankly speaking, I'm a novice in Rust, and it may take a good amount of time to finalize the test case. And, can I ask you for a favor to show me the best test case that I can refer to in [this file](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour/test.rs?rgh-link-date=2024-02-12T02%3A00%3A53Z)? I feel it can somehow speed up us a little bit. Thanks!

Unfortunately, this area of the code is also pretty foreign to me and I don't have the bandwidth to dig into it deeply.

File: protocols/kad/src/behaviour.rs [hunk 1]
```diff
@@ -2433,14 +2433,17 @@ where
         // Run the periodic provider announcement job.
         if let Some(mut job) = self.add_provider_job.take() {
             let num = usize::min(JOBS_MAX_NEW_QUERIES, jobs_query_capacity);
+            let mut ready_provider_keys: Vec<Key> = Vec::new();
             for _ in 0..num {
                 if let Poll::Ready(r) = job.poll(cx, &mut self.store, now) {
-                    self.start_add_provider(r.key, AddProviderContext::Republish)
```
Comment (thomaseizinger) [inline protocols/kad/src/behaviour.rs:2507]: Before we were calling `start_add_provider` for each one that was `Ready`, now we only do it for the first one. Isn't that wrong?

File: protocols/kad/src/behaviour.rs [hunk 2]
```diff
@@ -2433,14 +2433,17 @@ where
...
-                } else {
-                    break;
```
Comment (thomaseizinger) [inline protocols/kad/src/behaviour.rs:2509]: I think a better fix would be to move the `jobs_query_capacity -= ` statement before the break here but use the index of the `for` loop to decrease it by the right amount.

File: protocols/kad/src/behaviour.rs (hunk 1 above)
Comment (junekhan) [reply protocols/kad/src/behaviour.rs:2507]: You're right! Let me fix this.

**Review by thomaseizinger — DISMISSED** [review]
Cool! Happy to merge this without a test now.

Can you bump the version of the crate and add a changelog entry? See docs for details :)

File: protocols/kad/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,8 @@
+## 0.45.5
+
+- Compute `jobs_query_capacity` accurately.
+  See [PR 5148](https://github.com/libp2p/rust-libp2p/pull/5148).
+
 ## 0.45.4
```
Comment (thomaseizinger) [inline protocols/kad/CHANGELOG.md]: `0.45.4` isn't released yet. Add it to that section please!

**Review by jxs — DISMISSED** [review]
Thanks June for this, and Thomas for reviewing it! LGTM cc @iand

**Review by jxs — APPROVED** [review]

**Pull request merged**